core/src/main/java/org/infinispan/topology/ClusterCacheStatus.java

```diff
@@ -350,132 +336,160 @@ public void doConfirmRebalance(Address member, int receivedTopologyId) throws Ex
+      ConsistentHash newCurrentCH, newPendingCH = null;
+      CacheTopology.Phase newPhase = CacheTopology.Phase.STABLE;
+      List<Address> actualMembers;
+      if (newCurrentMembers.isEmpty()) {
+         // All the current members left, try to replace them with the joiners
+         log.tracef("All current members left, re-initializing status for cache %s", cacheName);
+         rebalanceConfirmationCollector = null;
 
-            newCurrentMembers = getExpectedMembers();
-            actualMembers = newCurrentMembers;
-            newCurrentCH = joinInfo.getConsistentHashFactory().create(
-                  joinInfo.getHashFunction(), joinInfo.getNumOwners(), joinInfo.getNumSegments(),
-                  newCurrentMembers, getCapacityFactors());
-         } else {
-            // ReplicatedConsistentHashFactory allocates segments to all its members, so we can't add any members here
-            newCurrentCH = consistentHashFactory.updateMembers(currentCH, newCurrentMembers, getCapacityFactors());
-            actualMembers = newCurrentMembers;
-            if (pendingCH != null) {
-               List<Address> newPendingMembers = pruneInvalidMembers(pendingCH.getMembers());
-               newPendingCH = consistentHashFactory.updateMembers(pendingCH, newPendingMembers, getCapacityFactors());
-               actualMembers = newPendingMembers;
-            }
+         newCurrentMembers = getExpectedMembers();
+         actualMembers = newCurrentMembers;
+         newCurrentCH = joinInfo.getConsistentHashFactory().create(
+               joinInfo.getHashFunction(), joinInfo.getNumOwners(), joinInfo.getNumSegments(),
+               newCurrentMembers, getCapacityFactors());
+      } else {
+         // ReplicatedConsistentHashFactory allocates segments to all its members, so we can't add any members here
+         newCurrentCH = consistentHashFactory.updateMembers(currentCH, newCurrentMembers, getCapacityFactors());
+         actualMembers = newCurrentMembers;
+         if (pendingCH != null) {
+            newPhase = currentTopology.getPhase();
+            newPendingCH = consistentHashFactory.updateMembers(pendingCH, newCurrentMembers, getCapacityFactors());
+            actualMembers = pruneInvalidMembers(pendingCH.getMembers());
          }
-         CacheTopology newTopology = new CacheTopology(topologyId + 1, rebalanceId, newCurrentCH, newPendingCH, actualMembers, persistentUUIDManager.mapAddresses(actualMembers));
-         setCurrentTopology(newTopology);
-
-         clusterTopologyManager.broadcastTopologyUpdate(cacheName, newTopology, availabilityMode,
-               isTotalOrder(), isDistributed());
       }
+      // Losing members during state transfer could lead to a state where we have more than two topologies
+      // concurrently in the cluster. We need to make sure that all the topologies are compatible (properties set
+      // in CacheTopology docs hold) - we just remove lost members.
```

Comment:
I think we need a new phase for the topology that has removed one member before starting the rebalance, otherwise there will be a mismatch between the meaning of `Phase.STABLE` and a stable topology as used by partition handling.